lib/puppet/pops/types/type_factory.rb

```diff
@@ -351,6 +351,22 @@ def self.hash_of_data()
     type
   end
 
+  # Produces a type for NotUndef[T]
+  # The given 'inst_type' can be a string in which case it will be converted into
+  # the type String[inst_type].
+  #
+  # @param inst_type [Type,String] the type to qualify
+  # @return [Puppet::Pops::Types::PNotUndefType] the NotUndef type
+  #
+  # @api public
+  #
+  def self.not_undef(inst_type = nil)
+    type = Types::PNotUndefType.new()
+    inst_type = string(inst_type) if inst_type.instance_of?(String)
```

Comment:
OK, but would anyone mock a String? The reason instance_of? is used here is to avoid is_a? (slow has hell in rgen).